MULE-19287 (DOCS-10601): Docs for Correlation ID (for v.4.4.0) (#1860)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/correlation-id.adoc

To change how Mule generates the correlation ID:

. Add the `<configuration>` component to your application XML.
. Set the `correlationIdGeneratorExpression` attribute to specify the expression that generates the correlation ID:
+
[source,xml,linenums]
----
<configuration correlationIdGeneratorExpression="#[<custom_generator_expression>]"/>
----

[IMPORTANT]
Ensure that the algorithm you use cannot generate the same correlation ID for two different events.

=== Correlation ID Generator Configuration Example

In the following example, Mule generates the correlation ID by calling the Java `uuid()` method, and then replaces the dashes (`-`) with dots (`.`):

[source,xml,linenums]
----
<configuration correlationIdGeneratorExpression="#[uuid() replace /-/ with('.')]"/>

<http:listener-config name="HTTP_Listener_config">
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>
<flow name="example">
  <http:listener config-ref="HTTP_Listener_config" path="/test"/>
  <set-payload value="some"/>
  <logger level="WARN" message="#[payload]"/>
</flow>
----

After executing the flow, the logger shows the following:

----

> WARN  2021-03-30 16:46:11,269 [[MuleRuntime].uber.06: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/1; event: b5a95a63.f190.4c91.880a.5b54512fa6b1] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some

Reviewer: Maybe in this section you can mention that the correlation ID for this example is b5a95a63.f190.4c91.880a.5b54512fa6b1 because this is the first time in the doc that shows an example of how an ID looks like.
Users might need to scroll down to the section _Example of Identifying the Correlation ID in Logs_ to have a first glance of a correlation "ID" example.

> = Correlation ID

Reviewer 2: Anything to add to this title? What _about_ the correlation ID is the focus of the topic? What it is? How to use it?

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: tracing, correlation, id, anypoint, studio

> The correlation ID is a unique string identifier for each Mule event. When Mule creates a new event, it also generates a correlation ID before sending the event to the next processor in the flow. This ID enables you to correlate different log entries to a particular execution. By default, Mule generates correlation IDs by generating a Java Universally Unique Identifier (UUID).

Reviewer 2: Just an idea to reduce words:
```suggestion
When Mule creates a new event, it generates a Java Universally Unique Identifier (UUID) string identifier called a *correlation ID* before sending the event to the next processor in the flow. This ID enables you to correlate different log entries with a particular execution. 
```

> The correlation ID can be used to understand the entire history of an event that resulted in an issue, or an unhandled error, because any error includes the correlation ID of the event that triggered it. Also, the correlation ID is very helpful while performing a memory analysis of heap dumps.

Reviewer 2: For active voice, try...:
```suggestion
Use the correlation ID to understand the entire history of an event that resulted in an issue or an unhandled error (all errors include the correlation ID of the event that triggered it). Also, knowing the correlation ID helps you perform a memory analysis of heap dumps.
```
Is it necessary to include "all errors include the correlation ID of the event that triggered it" if the user who gets the event sees it there?

> == Correlation ID Generator

Reviewer 2: The text seems more specifically about obtaining a correlation ID, since the first step is using an existing one rather than generating one....

Reviewer 2: I wonder if we can put Lines 13-16 in the intro topic, and Lines 18-21 to introduce ==Modify the..., thereby omitting one level of head and also putting all the text about moderation under one title?

> Mule runtime follows this procedure before generating a correlation ID:

Reviewer 2: This statement is a little misleading, implying a procedure _before_ generating the ID.
There is really only one step _before_ generating, and that is using the source ID.

So...along with moving it to the end of the previous container topic, I would reword it something like this:

"To obtain a correlation ID, Mule first checks for one in the source message (for example, an HTTP listener with the `X-CORRELATION-ID` header, or JMS). If the source does not set a correlation ID, Mule generates one using the correlation ID generator."

. If the source contains a correlation ID from the message (for example, an HTTP listener with the `X-CORRELATION-ID` header, or JMS), Mule uses that ID as the correlation ID of the event.

> . If the source does not set a correlation ID, Mule generates one.

Reviewer 2: See suggestions above.
Also, can you clarify how "or JMS" relates to "an HTTP listener with the `X-CORRELATION-ID` header": is JMS a substitute for the listener or the header?

> It is best to avoid making changes to the correlation ID generator, however, there are some scenarios where you might need to format or standardize the correlation ID for the events, for example:

Reviewer 2: So...where is the correlation ID generator described, or do we just mention it as the thing that creates the correlation ID if one does not exist in the source message?

Reviewer 2: ```suggestion
== Modify the Correlation ID Generator
Although it is best to avoid making changes to the correlation ID generator, you might need to format or standardize the correlation ID for the events if:
```

Reviewer 2: If you decide to go with this suggestion, remove "If" from the beginning of Lines 20 and 21.

> * If your company has its own standard or format for correlation IDs.

Reviewer 2: What is the difference between "standard" and "format"? 
Do we mean "standard format"?
Otherwise, I think "format" alone works.

> * If your external system differs in correlation ID lengths, causing incompatibility with Mule correlation IDs thus breaking traceability.

Reviewer 2: What does "your external system differs in correlation ID lengths" mean?
Differs from what?
Do we mean...:
```suggestion
* If your external system correlation IDs are a different length than those in Mule, causing an incompatibility that makes them untraceable. 
```
Just my best guess....

> === Modify the Correlation ID Generator

Reviewer 2: See reorg suggestion above.

To change how Mule generates the correlation ID:

. Add the `<configuration>` component to your application XML.

> . Set the `correlationIdGeneratorExpression` attribute and specify the expression that generates the correlation ID:

Reviewer 2: Is this two actions (set, specify) or one?
```suggestion
. Set the `correlationIdGeneratorExpression` attribute to specify the expression that generates the correlation ID:
```

+
[source,xml,linenums]
----
<configuration correlationIdGeneratorExpression="#[<custom_generator_expression>]"/>
----

[IMPORTANT]

> When you configure a correlation ID generator, ensure that the algorithm you use is not able to generate the same correlation ID for two different events.

Reviewer 2: ```suggestion
Ensure that the algorithm you use cannot generate the same correlation ID for two different events.
```

==== Configuration Example

In the following example, Mule generates the correlation ID by calling the Java `uuid()` method, and then replaces the dashes (`-`) with dots (`.`):

[source,xml,linenums]
----
<configuration correlationIdGeneratorExpression="#[uuid() replace /-/ with('.')]"/>

<http:listener-config name="HTTP_Listener_config">
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>
<flow name="example">
  <http:listener config-ref="HTTP_Listener_config" path="/test"/>
  <set-payload value="some"/>
  <logger level="WARN" message="#[payload]"/>
</flow>
----

After executing the flow, the logger shows the following:

----
WARN  2021-03-30 16:46:11,269 [[MuleRuntime].uber.06: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/1; event: b5a95a63.f190.4c91.880a.5b54512fa6b1] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some
----

> == Change the Correlation ID During the Flow Execution

Reviewer 2: Use the Mule Tracing Module to Change...?

> There are cases where you want to change the correlation ID of the event for a given scope or a set of operations, for example:

Reviewer 2: Again, all my suggestions are simply to get more directly to the point; your choice!
```suggestion
You might want to change the correlation ID of the event for a given scope or a set of operations if:
```

Reviewer 2: (Then remove "If" from bullets....)

Reviewer 2: "for a given scope": this scope is different from or the same as the Mule Tracing module `with-correlation-id` scope?

Reviewer 2: This intro sentence could also be...
```suggestion
Install the Mule Tracing module and use its `with-correlation-id` scope to change the correlation ID of the event for a given scope or a set of operations in the following circumstances:
```

> * If you are processing the results of a database query inside a `foreach` scope, and you want to correlate

Reviewer 2: ```suggestion
* If you are processing the results of a database query inside a `foreach` scope and you want to correlate
```

the event with the register you are processing.
* If you are consuming a JMS message queue and you want to proceed using the JMS correlation ID for traceability purposes.

> For these scenarios, the Mule Tracing module enables you to use the `with-correlation-id` scope to modify the correlation ID during the execution of said scope.

Reviewer 2: See suggestion above for working this in to the intro sentence.

Reviewer 2: Or...:
```suggestion
For these scenarios, installing the Mule Tracing module enables you to use the `with-correlation-id` scope to modify the correlation ID during the execution of said scope.
```

=== Install the Mule Tracing Module

> Follow the next steps to install the Mule Tracing module in your application.

Reviewer 2: ```suggestion
To install the Mule Tracing module in your application:
```

. Open your Mule project in Anypoint Studio.
. Go to the Mule Palette.

> . Select **Search in Exchange**, and search for the Mule Tracing module.

Reviewer 2: ```suggestion
. Select *Search in Exchange* and search for the Mule Tracing module.
```

  <set-payload value="some"/>
  <logger level="WARN" message="#[payload]"/>
  <tracing:with-correlation-id correlationId="#[correlationId ++ '-EXAMPLE']">
    <logger level="WARN" message="#[payload]"/>
  </tracing:with-correlation-id>
  <logger level="WARN" message="#[payload]"/>
</flow>
----

In this case, after the application executes the flow, the output logs are:
----
WARN  2021-03-30 16:46:11,269 [[MuleRuntime].uber.05: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/1; event: bad0e5b0-9191-11eb-a0b3-36548d51aeee] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some
WARN  2021-03-30 16:46:11,271 [[MuleRuntime].uber.05: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/2/processors/0; event: bad0e5b0-9191-11eb-a0b3-36548d51aeee-EXAMPLE] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some
WARN  2021-03-30 16:46:11,274 [[MuleRuntime].uber.05: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/3; event: bad0e5b0-9191-11eb-a0b3-36548d51aeee] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some
----

> === Correlation ID Logs in DEBUG Level

Reviewer 2: Use the DEBUG Log Level to View Additional Correlation ID Logs?

> In case you want to see additional logs related to the correlation ID, you can set the log level to `DEBUG` by adding the following code to your application XML:

Reviewer 2: ```suggestion
If you want to see additional logs related to the correlation ID, set the log level to `DEBUG` by adding the following code to your application XML:
```

----
<AsyncLogger name="org.mule.runtime.module.extension.internal.runtime.operation" level="DEBUG"/>
----

> After performing a request to the application from the previous example, the updated logs show as following:

Reviewer 2: ```suggestion
The updated logs display as follows:
```

WARN  2021-03-30 16:58:29,555 [[MuleRuntime].uber.05: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/2/processors/0; event: bad0e5b0-9191-11eb-a0b3-36548d51aeee-EXAMPLE] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some
DEBUG 2021-03-30 16:58:29,566 [[MuleRuntime].uber.05: [test-project-app].example.CPU_LITE @6d3b5ad] org.mule.runtime.module.extension.internal.runtime.operation.ImmutableProcessorChildContextChainExecutor: Event with correlationId '4c355b30-9192-11eb-b79f-36548d51aeee-EXAMPLE' going back to '4c355b30-9192-11eb-b79f-36548d51aeee' (successful execution) in location example/processors/2
WARN  2021-03-30 16:58:29,568 [[MuleRuntime].uber.05: [test-project-app].example.CPU_LITE @6d3b5ad] [processor: example/processors/3; event: bad0e5b0-9191-11eb-a0b3-36548d51aeee] org.mule.runtime.core.internal.processor.LoggerMessageProcessor: some
----

> == Correlation ID Scope and Error Handling

Reviewer 2: ```suggestion
== Error Handling
```

> In case of an error, the error handler logs the correlation ID that corresponds to the flow or Try scope that contains the error handler. For example:

Reviewer 2: ```suggestion
If an error occurs, the error handler logs the correlation ID that corresponds to the flow or Try scope that contains the error handler, for example:
```

[source,xml,linenums]
----
<flow name="example">
  <tracing:with-correlation-id correlationId="#[correlationId ++ '-EXAMPLE']">
    <raise-error type="APP:SOME"/>
  </tracing:with-correlation-id>
  <error-handler>
    <on-error-continue>
      <logger message="#[correlationId]"/>
    </on-error-continue>
  </error-handler>
</flow>
----

> In this scenario, Mule logs the correlation ID that the flow set at the beginning of the execution because the raised error stops the execution of the correlation ID scope, preventing the scope from setting the new correlation ID (`#[correlationId ++ '-EXAMPLE']`).

Reviewer 2: ```suggestion
In this scenario, Mule logs the correlation ID set at the beginning of the execution, because the error stops the execution of the correlation ID scope, preventing the it from setting the new correlation ID (`#[correlationId ++ '-EXAMPLE']`).
```

ERROR 2021-03-30 14:27:36,052 [[MuleRuntime].uber.08: [test-project-app].example.CPU_LITE @19afa17] [processor: example/processors/2; event: 37f0f591-917d-11eb-9209-3c22fb13cde7] org.mule.runtime.core.internal.exception.OnErrorPropagateHandler:
********************************************************************************
Message               : An error occurred.
Element               : example/processors/2 @ test-project-app:test-project-app.xml:16
Element DSL           : <raise-error type="APP:SOME"></raise-error>
Error type            : APP:SOME
FlowStack             : at example(example/processors/2 @ test-project-app:test-project-app.xml:16)

  (set debug level logging or '-Dmule.verbose.exceptions=true' for everything)
********************************************************************************
----

When the Logger and the Raise Error components execute, both components log the correlation ID `37f0f591-917d-11eb-9209-3c22fb13cde7`. Note that this correlation ID is different from the one logged in the previous example.

> == Example: Setting Correlation ID for a Database Query

Reviewer 2: ```suggestion
== Example of Setting the Correlation ID for a Database Query
```

> The following example shows a scenario where you can modify the correlation ID while processing the results of a database query:

Reviewer 2: ```suggestion
The following example shows a scenario in which you modify the correlation ID while processing the results of a database query:
```

image:mruntime-correlation-id-db-example.png[Correlation ID Database query example]

Application XML configuration:

[source,xml,linenums]
----
<flow name="correlation-id-flow">
  <db:select config-ref="Database_Config">
    <db:sql ><![CDATA[SELECT orderId, customerId, total from Orders  ]]></db:sql>
  </db:select>
  <foreach >
    <tracing:with-correlation-id correlationId="#[payload.orderId]">
      <logger level="INFO" message="#[correlationId]"/>
      <http:request method="GET" url="http://a-service.com" sendCorrelationId="ALWAYS"/>
    </tracing:with-correlation-id>
  </foreach>
</flow>
----

> In this case, each record obtained from the database is processed within the foreach scope, and the correlation ID for each iteration corresponds to the `orderId` of said record. In this scenario, if there is a problem during the HTTP request operation, the correlation ID enables you to track which record caused the error.

Reviewer 2: ```suggestion
In this case, each record obtained from the database is processed within the `foreach` scope, and the correlation ID for each iteration corresponds to the `orderId` of said record. In this scenario, if there is a problem during the HTTP request operation, the correlation ID enables you to track which record caused the error.
```
